elastic/logstash
# Fix the exception behavior when `config.string` contains ${VAR} in the comments.

## Release notes

[rn:skip]

## What does this PR do?
We were applying var substitution once we load from YAML and it affects when we use `config.string`. Other options like `path.config` since points to path, the configs will be [loaded later](https://github.com/mashhurs/logstash/blob/main/logstash-core/lib/logstash/config/source/local.rb#L202-L216) and LSCL rules will be processed when compiling the pipeline. This is _why_ we are experiencing the issue with `config.string` _only_.

This PR standardizes (same sequence will be with `config.string` as well) the way we are loading configs and compiling.

## Why is it important/What is the impact to the user?
Probability might be low but it is still not an expectation to resolve the `${VAR}` if commented out. So, users may have a better experience with this change.

## Checklist


## Author's Checklist


## How to test this PR locally

- pull the change
- use following config in the `pipelines.yml`
  ```
   - pipeline.id: test
  config.string: |
    input { generator {} } #${I_AM_NOT_HERE}
    filter { sleep { time => 100 } }
    output {
    #${I_AM_NOT_HERE}
      stdout {
        codec => rubydebug
      }
      # ${I_AM_NOT_HERE}
    }
  ```
- Run Logstash with `bin/logstash --debug --config.debug` to see the what config was applied. 
- Logstash runs without crash. Before the change, Logstash was failing `Cannot evaluate ${I_AM_NOT_HERE}... error`
  - and config will be applied after filtering out `${I_AM_NOT_HERE}` with the comments
    ```
     ----------------------P[input-generator{"count"=>5}|[string]config_string:1:9:```
     generator { count => 5 }
     ```] -> __QUEUE__
     __QUEUE__ -> filter_to_output
     filter_to_output -> P[output-stdout{"codec"=>#<Java::OrgLogstashConfigIrImperative::PluginStatement:0x3344b612>}|     
     [string]config_string:3:1:```
     stdout {
     codec => rubydebug 
     }
     ```]
    ```

## Related issues

- Closes #16008 

## Use cases

## Screenshots

## Logs
